### PR TITLE
fix: Flush stdout and stderr on each write when debugging

### PR DESCRIPTION
**Pull request by ndobryanskyy**
*Issue #, if available:*
#835 
*Description of changes:*

Introduced `StreamWriter` which wraps `stdout` and `stderr` and is passed to the Lambda container. This class supports `auto_flushing` and could be further extended to support any flexibility in output as we want.

*Checklist:*

- [x] Write unit tests
- [x] Write/update functional tests. Do I need any?
- [x] Write/update integration tests. Do I need any?
- [x] `make pr` passes
- [x] Write documentation. Updated all existing doc strings referring to `stdout` and `stderr`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: samcli/lib/utils/stream_writer.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
+"""
+This class acts like a wrapper around output streams to provide any flexibility with output we need
+"""
```
Comment (TheSriram) [inline samcli/lib/utils/stream_writer.py]: I'm curious to know what is the performance hit, if we just flush all the time.

**Comment by gyfis** [comment]
While I don't have much context about the overall direction of the sam cli, I really like these changes and the fact that it's not just an if-clause somewhere, but a more complete approach to the problem. Thank you for taking the time to work on this :)

File: samcli/lib/utils/stream_writer.py (hunk 1 above)
Comment (ndobryanskyy) [reply samcli/lib/utils/stream_writer.py]: I think it is completely negligible for the debugging, thoughts?

Comment (TheSriram) [reply samcli/lib/utils/stream_writer.py]: yeah totally we need it for debugging. I'm talking about non-debugging cases as well. just curious.

**Comment by ndobryanskyy** [comment]
@TheSriram I wonder though why on python3.7 integration tests fail?

**Comment by jfuss** [comment]
The Golang zip runtime test is being funky on 3.6 and 3.7 as of recently. I just restarted it. We aren't entirely sure why it fails sometimes at the moment.

**Comment by ndobryanskyy** [comment]
@jfuss, understood, pretty strange 😅. Thanks for restarting tho, now it passed just fine :D

**Comment by ndobryanskyy** [comment]
@TheSriram, thanks for the approve, the only thing that concerns me is that should we really think about "flushing all three time" as it is actually requested in #835 
Because my PR actually fixes only cases intended for debugging, thoughts?

**Comment by TheSriram** [comment]
Correct, that is how I'm thinking of this as well. This PR adds mechanisms around how we flush. But flushing all the time should definitely be a separate PR, which validates performance and what are the impacts of such an operation.

**Comment by ndobryanskyy** [comment]
Definitely, also I'm looking forward to investigate unbuffered streams in Python3 (in Python2 they just work), that should be a good thing (performancewise) if we want "auto flush" all the time behavior

**Pull request merged**